File: C++/stack/stack_array.cpp
//
// Created by Franc on 10/11/2025.
//

#include <cstddef>
#include <memory>
#include <optional>
#include <iostream>

class Stack
{
    static constexpr std::size_t EMPTY_STACK = -1;

    std::unique_ptr<int[]> stack;
    std::size_t top = -1;
    const std::size_t capacity;

public:
    explicit Stack(const std::size_t capacity)
        : capacity(capacity)
    {
        stack = std::make_unique<int[]>(capacity);
    }

    bool push(const int val)
    {
        if (top == capacity - 1)
        {
            std::cout << "La pila esta llena." << std::endl;
            return false;
        }

        stack[++top] = val;
        return true;
    }

    std::optional<int> pop()
    {
        if (top == EMPTY_STACK)
        {
            std::cout << "La pila esta vacia." << std::endl;
            return std::nullopt;
        }


        return stack[top--];
    }

    [[nodiscard]] std::optional<int> peek() const
    {
        if (top == EMPTY_STACK)
        {
            std::cout << "La pila esta vacia." << std::endl;
            return std::nullopt;
        }

        return stack[top];
    }

    [[nodiscard]] bool isEmpty() const
    {
        return top == EMPTY_STACK;
    }

    [[nodiscard]] bool isFull() const
    {
        return top == capacity - 1;
    }
};

int main()
{
    auto stack = Stack(10);
    stack.push(10);
    stack.push(20);
    stack.push(30);

    std::cout << "Elemento superior: " << stack.pop().value() << std::endl;
    std::cout << "Elemento superior (sin extraer): " << stack.peek().value() << std::endl;
    std::cout << "Elemento superior (extraido): " << stack.pop().value() << std::endl;
    std::cout << "Ultimo elemento: " << stack.pop().value() << std::endl;

    return 0;
}

File: Python/stack/stack_array.py
class Stack:
    def __init__(self, capacity):
        self.capacity = capacity
        self.stack = [0] * capacity
        self.top = -1

    def push(self, value):
        if self.isFull():
            print("La pila esta llena.")
            return False

        self.top += 1
        self.stack[self.top] = value
        return True

    def pop(self):
        if self.isEmpty():
            print("La pila esta vacia.")
            return -1

        val = self.stack[self.top]
        self.top -= 1
        return val

    def peek(self):
        if self.isEmpty():
            print("La pila esta vacia.")
            return -1

        return self.stack[self.top]

    def isEmpty(self):
        return self.top == -1

    def isFull(self):
        return self.top == self.capacity - 1


stack = Stack(5)
stack.push(10)
stack.push(20)
stack.push(30)

print("Elemento superior: " + str(stack.pop()))
print("Elemento superior (sin extraer): " + str(stack.peek()))
print("Elemento superior (extraido): " + str(stack.pop()))
print("Ultimo elemento: " + str(stack.pop()))
